File: petclinic-java-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/test/java/dev/pollito/petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts/LoginSpec.java
package dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts;

import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.config.BaseIntegrationSpec;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class LoginSpec extends BaseIntegrationSpec {

    private static final String USERNAME = "admin";
    private static final String PASSWORD = "admin123";

    @BeforeEach
    void setUp() {
        initMockMvc();
        createTestUser(USERNAME, PASSWORD);
    }

    @Test
    void shouldRedirectToLoginWhenAccessingRoot() throws Exception {
        mockMvc.perform(get("/"))
                .andExpect(status().is3xxRedirection());
    }

    @Test
    void shouldLoginWithValidCredentials() throws Exception {
        mockMvc.perform(post("/login")
                .with(csrf())
                .param("username", USERNAME)
                .param("password", PASSWORD))
                .andExpect(status().is3xxRedirection());
    }

    @Test
    void shouldFailLoginWithInvalidCredentials() throws Exception {
        mockMvc.perform(post("/login")
                .with(csrf())
                .param("username", USERNAME)
                .param("password", "wrongpassword"))
                .andExpect(status().is3xxRedirection());
    }
}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/test/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/LoginSpec.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.config.BaseIntegrationSpec
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

class LoginSpec : BaseIntegrationSpec() {

    companion object {
        private const val USERNAME = "admin"
        private const val PASSWORD = "admin123"
    }

    @BeforeEach
    fun setUp() {
        initMockMvc()
        createTestUser(USERNAME, PASSWORD)
    }

    @Test
    fun shouldRedirectToLoginWhenAccessingRoot() {
        mockMvc.perform(MockMvcRequestBuilders.get("/"))
            .andExpect(MockMvcResultMatchers.status().is3xxRedirection())
    }

    @Test
    fun shouldLoginWithValidCredentials() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/login")
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .param("username", USERNAME)
                .param("password", PASSWORD),
        )
            .andExpect(MockMvcResultMatchers.status().is3xxRedirection())
    }

    @Test
    fun shouldFailLoginWithInvalidCredentials() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/login")
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .param("username", USERNAME)
                .param("password", "wrongpassword"),
        )
            .andExpect(MockMvcResultMatchers.status().is3xxRedirection())
    }
}
